front/pages/w/[wId]/u/chat/[cId]/index.tsx

```diff
@@ -825,20 +902,25 @@ export default function AppChat({
     }
     throw new Error("Error: no OUTPUT block streamed.");
   };
-  const updateMessages = (
+  const updateMessages = async (
     messages: ChatMessageType[],
-    userMessage: ChatMessageType
-  ): void => {
-    messages.push(userMessage);
+    newMessage: ChatMessageType
+  ): Promise<void> => {
+    if (!newMessage.sId) newMessage.sId = client_side_new_id();
+    if (newMessage.role !== "error") await upsertNewMessage(newMessage);
+    messages.push(newMessage);
     setMessages(messages);
     setResponse(null);
   };
 
-  function filterErrorMessages(m: ChatMessageType[]): void {
+  async function filterErrorMessages(m: ChatMessageType[]): Promise<void> {
     // remove last message if it's an error, and the previous messages until the
     // last user message, included
     if (m.length === 0 || m[m.length - 1].role !== "error") return;
-    while (m.pop()?.role !== "user" && m.length > 0);
+    let message = m.pop(); // remove error message which has not been stored
+    while (m.length > 0 && (message = m.pop())?.role !== "user")
+      await deleteMessage(message as ChatMessageType); // remove messages until last user message
```

Comment:
Why are we upserting error messages in the first place? Does it even work? It seems weird to delete while we filter at each submit.

Reply:
error messages are not upserted so not deleted (deleteMessage is not called on the error message here)
for history sanity we remove messages up to the user message that triggered the error; we cannot know them in advance unless complicating the logic a bit

I understand the filtering feels weird, I'll look for a better solution

Reply:
Yes that's quite confusing to the reader if we never call deleteMessage but do write the code there :+1:

Reply:
As per IRL this is the current behavior so keeping the same logic for now :+1: